### PR TITLE
Initial PR

### PR DESCRIPTION
**Pull request by fleetwoods**
WIP......

### CODE REVIEW

File: project/Dependencies.scala [hunk 1]
```diff
@@ -1,13 +1,21 @@
...
+    "com.typesafe.akka"   %% "akka-http-core"            % "10.0.0",
+    "com.typesafe.akka"   %% "akka-stream-kafka"         % "0.12",
+    "net.cakesolutions"   %% "scala-kafka-client"        % "0.10.0.0",
+    "com.ovoenergy"       %% "comms-kafka-messages"      % kafkaMessagesVersion,
+    "com.ovoenergy"       %% "comms-kafka-serialisation" % kafkaMessagesVersion,
+    "com.typesafe.akka"   %% "akka-slf4j"                % "2.3.14",
+    "ch.qos.logback"       % "logback-classic"           % "1.1.7",
+    "me.moocar"            % "logback-gelf"              % "0.2",
+    "io.logz.logback"      % "logzio-logback-appender"   % "1.0.11",
+    "org.typelevel"       %% "cats-free"                 % "0.8.1",
```
Comment (cb372) [inline project/Dependencies.scala]: You only need `cats-core` in this app, not `cats-free`

File: src/main/scala/com/ovoenergy/orchestration/kafka/Serialisation.scala [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package com.ovoenergy.orchestration.kafka
+
+import com.ovoenergy.comms.model.{OrchestratedEmail, Triggered}
+import com.ovoenergy.comms.serialisation.Serialisation._
+
+object Serialisation {
```
Comment (cb372) [inline src/main/scala/com/ovoenergy/orchestration/kafka/Serialisation.scala]: Not sure this object is really necessary, as each val is only used once. You could just as easily call `avroDeserializer[Triggered]` in `Main.scala`.

File: src/main/scala/com/ovoenergy/orchestration/kafka/OrchestrationFlow.scala [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import com.ovoenergy.comms.model.{CustomerProfile, Triggered}
+import com.ovoenergy.orchestration.logging.LoggingWithMDC
+import org.apache.kafka.common.serialization.{Deserializer, StringDeserializer}
+
+import scala.concurrent.Future
+import scala.util.control.NonFatal
+
+case class OrchestrationFlowConfig(hosts: String, groupId: String, topic: String)
+
+object OrchestrationFlow extends LoggingWithMDC {
```
Comment (cb372) [inline src/main/scala/com/ovoenergy/orchestration/kafka/OrchestrationFlow.scala]: What we're building here is actually an akka-streams `Graph`, not a `Flow`. Maybe the object should be renamed to reflect that.

File: src/main/scala/com/ovoenergy/orchestration/processes/Orchestrator.scala [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+object Orchestrator {
+
+  def apply(emailOrchestrator: (CustomerProfile, Triggered) => Future[_])(triggered: Triggered) = {
+
+    def determineOrchestrator(channel: Channel): (CustomerProfile, Triggered) => Future[_] = {
+      channel match {
+        case Email =>
+          emailOrchestrator
+        case _ =>
+          (customerProfile: CustomerProfile, triggered: Triggered) =>
```
Comment (cb372) [inline src/main/scala/com/ovoenergy/orchestration/processes/Orchestrator.scala]: Just for readability I'd make this anonymous function a field with a descriptive name e.g. `val logErrorAndSkip`

File: src/main/scala/com/ovoenergy/orchestration/processes/Orchestrator.scala [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+      }
+    }
+
+    val customerProfile = CustomerProfiler(triggered.metadata.customerId)
+    val orchestrator = customerProfile.map(ChannelSelector(_)).map(determineOrchestrator(_))
+
+    for {
+      customerProfile <- customerProfile
+      orchestrator <- orchestrator
+    } yield orchestrator(customerProfile, triggered)
```
Comment (cb372) [inline src/main/scala/com/ovoenergy/orchestration/processes/Orchestrator.scala]: I found this bit of code a bit of confusing for a couple of reasons:

1. Mapping over `customerProfile` and then using the result in a for-comprehension (which is actually another map)
2. Shadowing the name in the right-hand side with the left-hand side in the for-comprehension

I would probably write it something like this (may or may not compile):

```scala
val customerProfile = CustomerProfiler(triggered.metadata.customerId)
for {
  cp <- customerProfile
  channel <- ChannelSelector(cp)
  orchestrator <- determineOrchestrator(channel)
} yield orchestrator(cp, triggered)
```

File: src/main/scala/com/ovoenergy/orchestration/processes/email/EmailOrchestration.scala [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+      else Validated.valid(customerProfile.name.firstName)
+    }
+
+    val lastName: ValidationErrorsOr[String] = {
+      if (customerProfile.name.lastName.isEmpty) Validated.invalid(ValidationErrors("Customer has no last name"))
+      else Validated.valid(customerProfile.name.lastName)
+    }
+
+    val resultOrValidationErrors: ValidationErrorsOr[Future[_]] =
+      Apply[ValidationErrorsOr].map3(emailAddress, firstName, lastName) {
```
Comment (cb372) [inline src/main/scala/com/ovoenergy/orchestration/processes/email/EmailOrchestration.scala]: Check out your cats skillz! Nice.

File: src/main/scala/com/ovoenergy/orchestration/kafka/Serialisation.scala (hunk 2 above)
Comment (fleetwoods) [inline src/main/scala/com/ovoenergy/orchestration/kafka/Serialisation.scala]: These are used to various locations (but yes only once), service tests and application code, will revisit later.

File: src/main/scala/com/ovoenergy/orchestration/kafka/OrchestrationGraph.scala [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+    Consumer
+      .committableSource(consumerSettings, Subscriptions.topics(config.topic))
+      .mapAsync(1)(msg => {
+        log.debug(s"Event received $msg")
+        val result = msg.record.value match {
+          case Some(triggered) =>
+            orchestrationProcess(triggered).recover({
+              case NonFatal(error) =>
+                //TODO - Raise failed event
```
Comment (cb372) [inline src/main/scala/com/ovoenergy/orchestration/kafka/OrchestrationGraph.scala]: TODO

**Comment by fleetwoods** [comment]
TODOs in next PR, this is getting too big

**Comment by ovo-comms-circleci** [comment]
This PR was successfully deployed to production by [CircleCI build 13](https://circleci.com/gh/ovotech/comms-orchestration/13) at 2016-12-14 16:24 GMT.

**Pull request merged**